ovs/extensions/packages/debian.py

```diff
@@ -29,112 +28,108 @@ class DebianPackage(object):
-                installed = Toolbox.remove_prefix(line, 'Installed:').strip()
-            elif line.startswith('Candidate:'):
-                candidate = Toolbox.remove_prefix(line, 'Candidate:').strip()
-            if installed is not None and candidate is not None:
-                break
-        return installed if installed != '(none)' else None, candidate if candidate != '(none)' else None
-
-    @staticmethod
-    def get_versions(client):
+    def get_installed_versions(client=None, package_names=None):
+        """
+        Retrieve currently installed versions of all packages
+        :param client: Client on which to check the installed versions
+        :type client: SSHClient
+        :param package_names: Name of the packages to check
+        :type package_names: list
+        :return: Package installed versions
+        :rtype: dict
+        """
         versions = {}
-        for package_name in DebianPackage.OVS_PACKAGE_NAMES:
-            version_info = DebianPackage._get_version(package_name, client)
-            if version_info:
-                versions[package_name] = version_info
+        if package_names is None:
+            package_names = DebianPackage.OVS_PACKAGE_NAMES
+        for package_name in package_names:
+            command = "dpkg -s {0} | grep Version | awk '{{print $2}}'".format(package_name)
+            if client is None:
+                versions[package_name] = check_output(command, shell=True).strip()
+            else:
+                versions[package_name] = client.run(command, allow_insecure=True).strip()
         return versions
 
     @staticmethod
-    def install(package_name, client, force=False):
-        force_text = '--force-yes' if force is True else ''
-        counter = 0
-        max_counter = 5
-        last_exception = None
-        success = False
-        while counter < max_counter:
-            counter += 1
-            try_again = ', trying again' if counter < max_counter else ''
-            try:
-                client.run(['apt-get', 'install', '-y', force_text, package_name])
-                installed, candidate = DebianPackage._get_installed_candidate_version(package_name, client=client)
-                if installed == candidate:
-                    success = True
+    def get_candidate_versions(client, package_names):
+        """
+        Retrieve the versions candidate for installation of all packages
+        :param client: Root client on which to check the candidate versions
+        :type client: SSHClient
+        :param package_names: Name of the packages to check
+        :type package_names: list
+        :return: Package candidate versions
+        :rtype: dict
+        """
+        DebianPackage.update(client=client)
+        versions = {}
+        for package_name in package_names:
+            versions[package_name] = ''
+            for line in client.run(['apt-cache', 'policy', package_name, DebianPackage.APT_CONFIG_STRING]).splitlines():
```

Comment:
Is it possible that we have these dpkg lock issues here? If so, can we implement some re-try logic to wait for X minutes trying again every now and then?